scene/resources/curve.cpp

```diff
@@ -1496,76 +1499,138 @@ void Curve3D::_bake() const {
 		Vector3 npp = points[i + 1].position;
 		real_t d = position.distance_to(npp);
 
-		position = npp;
-		Plane post;
-		post.normal = position;
-		post.d = points[i + 1].tilt;
+		if (d > CMP_EPSILON) { // avoid the degenerate case of two very close points
+			position = npp;
+			Plane post;
+			post.normal = position;
+			post.d = points[i + 1].tilt;
 
-		dist += d;
+			dist += d;
 
-		pointlist.push_back(post);
-		distlist.push_back(dist);
+			pointlist.push_back(post);
+			distlist.push_back(dist);
+		}
 	}
 
 	baked_max_ofs = dist;
 
-	baked_point_cache.resize(pointlist.size());
-	Vector3 *w = baked_point_cache.ptrw();
-	int idx = 0;
+	const int point_count = pointlist.size();
+	{
+		baked_point_cache.resize(point_count);
+		Vector3 *w = baked_point_cache.ptrw();
 
-	baked_tilt_cache.resize(pointlist.size());
-	real_t *wt = baked_tilt_cache.ptrw();
+		baked_tilt_cache.resize(point_count);
+		real_t *wt = baked_tilt_cache.ptrw();
 
-	baked_up_vector_cache.resize(up_vector_enabled ? pointlist.size() : 0);
-	Vector3 *up_write = baked_up_vector_cache.ptrw();
+		baked_dist_cache.resize(point_count);
+		real_t *wd = baked_dist_cache.ptrw();
 
-	baked_dist_cache.resize(pointlist.size());
-	real_t *wd = baked_dist_cache.ptrw();
+		int idx = 0;
+		for (const Plane &E : pointlist) {
+			w[idx] = E.normal;
+			wt[idx] = E.d;
+			wd[idx] = distlist[idx];
 
-	Vector3 sideways;
-	Vector3 up;
-	Vector3 forward;
+			idx++;
+		}
+	}
+
+	if (!up_vector_enabled) {
+		baked_up_vector_cache.resize(0);
+		return;
+	}
 
-	Vector3 prev_sideways = Vector3(1, 0, 0);
-	Vector3 prev_up = Vector3(0, 1, 0);
-	Vector3 prev_forward = Vector3(0, 0, 1);
+	// Step 2: Caculate up vectors and whole local reference frames
+	//
+	// see C. Dougan, The Parallel Transport Frame, Game Programming Gems 2 for example
+	// for a discussion about why not Frenet frame.
+	{
+		PackedVector3Array baked_forward_vector_cache;
```

Comment:
This array is obviously not cached, so this variable name is weird.

It should be renamed something like `forward_vectors` or `forward_vector_list` instead of `baked_forward_vector_cache`.

Reply:
done. changed to forward_vectors